### PR TITLE
New magic %%executable to compile current AST

### PR DESCRIPTION
**Pull request by hahnjo**
The cell content is used as the `main` function. After declaration, use a `RecursiveASTVisitor` to generate LLVM IR for all currently parsed `FunctionDecl`s and `VarDecl`s at file scope in the AST. Then emit object code to a temporary file and finally invoke the linker to produce a binary that can be execute with the `!` magic command.

### CODE REVIEW

**Comment by hahnjo** [comment]
This very likely needs user documentation in `docs/source/magics.rst`, but I first want to check if you would be in favor of including a feature like this. We have found a ton of use cases, starting from using ThreadSanitizer (which would be very tricky in a JIT context) over more control in a benchmark environment to possibly using MPI.

I have a commit locally that hooks up things needed for `-fsanitize=thread`. Currently it just checks if the argument is part of the linker arguments (string comparison, so no real parsing), but I can include it if you want.

File: CMakeLists.txt [hunk 1]
```diff
@@ -167,6 +167,8 @@ set(XEUS_CLING_SRC
     src/xpaths.cpp
     src/xpaths.hpp
     src/xholder_cling.cpp
+    src/xmagics/executable.cpp
+    src/xmagics/executable.hpp
```
Comment (SylvainCorlay) [inline CMakeLists.txt]: (typo) the file is repeated.

Comment (hahnjo) [reply CMakeLists.txt]: No, the first is the source file `.cpp` and the second one the header `.hpp`. I made this the same as existing code for other magics.

Comment (SylvainCorlay) [reply CMakeLists.txt]: indeed, I was going too quickly over the PR.

**Comment by SylvainCorlay** [comment]
Thanks @hahnjo and sorry for the late reply!

I would be very happy to include this. Do you want to include some documentation before we move forward and merge this?

**Comment by hahnjo** [comment]
> Thanks @hahnjo and sorry for the late reply!
> 
> I would be very happy to include this. Do you want to include some documentation before we move forward and merge this?

Great! Yes, I'll add some user documentation on Monday.

**Comment by hahnjo** [comment]
@SylvainCorlay PTAL

**Comment by SylvainCorlay** [comment]
@hahnjo I am curious about the behavior of the resulting binary when you make use of functions from the kernel (to display things in the notebook etc).

What would be really awesome would be for the binary to actually link with xeus :smile:

**Comment by hahnjo** [comment]
> @hahnjo I am curious about the behavior of the resulting binary when you make use of functions from the kernel (to display things in the notebook etc).
> 
> What would be really awesome would be for the binary to actually link with xeus smile

Not sure what the expected result would be: You're running the binary as a new process and Jupyter just forwards the output streams. So there is no (direct) connection to the notebook that you could use for communication.

**Comment by SylvainCorlay** [comment]
More generally, can the result be linking with other binaries?

**Comment by SylvainCorlay** [comment]
Where I am going with this - which is kindof a moon shot and probably out of scope for this PR is the folllowing:

We have this project called "Voilà" which turns notebooks into standalone web apps. In the case of the C++ kernel, since the content of the notebook is fixed, there is no real need for the C++ to be "interpreted", and we could actually turn the  "notebook content + kernel" into a single executable that speaks to the voilà front-end. By all practical purposes, the resulting executable would be "a kernel" that only executes the content of the notebook and responds to comms.

**Comment by hahnjo** [comment]
> More generally, can the result be linking with other binaries?

Yes, that is actually the reason for passing the additional arguments of `%%executable` to the linker. I've successfully used this for OpenMP (`-fopenmp`), TSan (`-fsanitize=thread`; needs some tweaks for the CodeGenOptions), and MPI (`-lmpi`).

**Comment by SylvainCorlay** [comment]
cc @JohanMabille

**Comment by hahnjo** [comment]
I've rebased this PR on top of current `master` and also reformatted `if` statements (sorry for not noticing earlier). The new version also explicitly handles `FunctionDecl`s and global `VarDecl`s instead of relying on `HandleTopLevelDecl` to ignore "invalid" decls: The old code triggered assertions when enabled.

**Comment by hahnjo** [comment]
@SylvainCorlay @JohanMabille any update on merging this? I've also cleaned up my implementation of rudimentary option parsing for `-fsanitize=thread`, but I think this should be a separate PR...

**Comment by SylvainCorlay** [comment]
Sorry I actually thought that I had merged. This would have gotten into the next release anyways :)

Many thanks for this!

**Pull request merged**